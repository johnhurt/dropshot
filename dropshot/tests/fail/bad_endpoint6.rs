// Copyright 2020 Oxide Computer Company

#![allow(unused_imports)]

use dropshot::endpoint;
use dropshot::HttpError;
use dropshot::HttpResponseOk;
use dropshot::RequestContext;
use schemars::JsonSchema;


#[derive(JsonSchema)]
#[allow(dead_code)]
struct Ret {
    x: String,
    y: u32,
}

#[endpoint {
    method = GET,
    path = "/test",
}]
async fn bad_endpoint(_rqctx: &mut RequestContext<()>) -> Result<HttpResponseOk<Ret>, HttpError> {
    Ok(HttpResponseOk(Ret { "Oxide".to_string(), 0x1de }))
}

fn main() {}
