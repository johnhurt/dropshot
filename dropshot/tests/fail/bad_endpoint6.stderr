error[E0277]: the trait bound `Ret: serde::ser::Serialize` is not satisfied
  --> $DIR/bad_endpoint6.rs:23:23
   |
23 |     Ok(HttpResponseOk(Ret { x: "Oxide".to_string(), y: 0x1de }))
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `serde::ser::Serialize` is not implemented for `Ret`
   |
   = note: required by `HttpResponseOk`

error[E0277]: the trait bound `Ret: serde::ser::Serialize` is not satisfied
    --> $DIR/bad_endpoint6.rs:23:8
     |
23   |     Ok(HttpResponseOk(Ret { x: "Oxide".to_string(), y: 0x1de }))
     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `serde::ser::Serialize` is not implemented for `Ret`
     |
    ::: $WORKSPACE/dropshot/src/handler.rs
     |
     | pub struct HttpResponseOk<T: JsonSchema + Serialize + Send + Sync + 'static>(
     |                                           --------- required by this bound in `HttpResponseOk`

error[E0277]: the trait bound `Ret: serde::ser::Serialize` is not satisfied
    --> $DIR/bad_endpoint6.rs:23:5
     |
23   |     Ok(HttpResponseOk(Ret { x: "Oxide".to_string(), y: 0x1de }))
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `serde::ser::Serialize` is not implemented for `Ret`
     |
    ::: $WORKSPACE/dropshot/src/handler.rs
     |
     | pub struct HttpResponseOk<T: JsonSchema + Serialize + Send + Sync + 'static>(
     |                                           --------- required by this bound in `HttpResponseOk`

error[E0277]: the trait bound `Ret: serde::ser::Serialize` is not satisfied
    --> $DIR/bad_endpoint6.rs:22:94
     |
22   |   async fn bad_endpoint(_rqctx: &RequestContext<()>) -> Result<HttpResponseOk<Ret>, HttpError> {
     |  ______________________________________________________________________________________________^
23   | |     Ok(HttpResponseOk(Ret { x: "Oxide".to_string(), y: 0x1de }))
24   | | }
     | |_^ the trait `serde::ser::Serialize` is not implemented for `Ret`
     |
    ::: $WORKSPACE/dropshot/src/handler.rs
     |
     |   pub struct HttpResponseOk<T: JsonSchema + Serialize + Send + Sync + 'static>(
     |                                             --------- required by this bound in `HttpResponseOk`

error[E0277]: the trait bound `Ret: serde::ser::Serialize` is not satisfied
    --> $DIR/bad_endpoint6.rs:18:1
     |
18   | / #[endpoint {
19   | |     method = GET,
20   | |     path = "/test",
21   | | }]
     | |__^ the trait `serde::ser::Serialize` is not implemented for `Ret`
     |
    ::: $WORKSPACE/dropshot/src/handler.rs
     |
     |   pub struct HttpResponseOk<T: JsonSchema + Serialize + Send + Sync + 'static>(
     |                                             --------- required by this bound in `HttpResponseOk`

error[E0277]: the trait bound `fn(RequestContext<()>) -> impl Future {wrapped_bad_endpoint}: dropshot::handler::HttpHandlerFunc<_, _, _>` is not satisfied
  --> $DIR/bad_endpoint6.rs:18:1
   |
18 | / #[endpoint {
19 | |     method = GET,
20 | |     path = "/test",
21 | | }]
   | |__^ the trait `dropshot::handler::HttpHandlerFunc<_, _, _>` is not implemented for `fn(RequestContext<()>) -> impl Future {wrapped_bad_endpoint}`
   |
  ::: $WORKSPACE/dropshot/src/api_description.rs
   |
   |           FuncParams: Extractor + 'static,
   |                                   ------- required by this bound in `ApiEndpoint::<Context>::new`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
